### ISSUE TITLE
Raise does not work with KVM 

### ISSUE THREAD

**Issue by Nuru** [Dygmalab/Bazecor#157]
**Describe the bug**
(First reported by email a year ago, Dec 31, 2019)

I am using an IOGEAR GCS1644 KVM, and the Raise is not compatible with it as the controlling keyboard. This is a big problem for me.

The controlling keyboard on the KVM is special in that the KVM intercepts certain keystrokes used to control the KVM itself. For example, to switch from monitoring port 1 to port 4, I can type `ctrl ctrl 4 enter`. It is also special in that when control is switched away from a computer, the KVM continues to supply an emulated keyboard to the computer, so the computer does not think the keyboard is being unplugged. It is the only satisfactory way I have of using a keyboard with my set of computers.

I do not expect Bazecor to be able to do a firmware update in this configuration, but I do expect that I can use the Raise as a normal keyboard. 

**To Reproduce**
Plug the Raise into the keyboard port of the KVM

**Expected behavior**
The keyboard works like an ordinary keyboard

**Actual behavior**
The Neuron flickers on for a flash every few seconds and the keyboard never lights up or is recognized

**Desktop (please complete the following information):**
 - OS: MacOS Catalina
 - Bazecor Version: 0.2.6

**Additional context**
The keyboard works as expected when connected directly to a computer.

On January 2, 2020, I received email from Dygma with this explanation
> It's a 6KRO vs NKRO issue. Most KVM switches only recognize 6KRO, but do not do what the USB spec would require them to do: ask the keyboard to go 6KRO. They just assume it will be. There's no reliable way to automatically switch to 6KRO, unless the host (in this case, the KVM) asks us to do so.
>
>With the next firmware update, there will be the 6KRO<->NKRO toggle, which will help with this issue. Press the combo, the keyboard goes 6KRO, KVM switch will recognize it, all's well! 

That was several releases ago (I think even before Bazecor 0.2.0), but I have yet to see this toggle feature announced or documented. Has it been implemented? Was it tried and found not to be a solution?

**Comment by AlexDygma** [comment]
Hey there @Nuru 

I'm sorry to hear you have had this problem for so long.

If the raise flickers this maybe due to lack of current being driven through the USB, the rase needs nearly 1 amp to work with all leds lit up, try to plug just half of the raise first (neuron and left side for example) if it work there is a chance that just by dimming the LED's you can use it, in extreme cases you can switch them off completely.

test it and lets see how it goes!

**Comment by Nuru** [comment]
@AlexDygma wrote
> If the raise flickers this maybe due to lack of current being driven through the USB, the rase needs nearly 1 amp to work with all leds lit up, try to plug just half of the raise first (neuron and left side for example) if it work there is a chance that just by dimming the LED's you can use it, in extreme cases you can switch them off completely.

Not a bad conjecture, but it does not seem to be correct. The KVM has a USB 2.0 auxiliary port. I happen to have a USB ammeter, and when I connect half the board to the auxiliary port, it works fine, and the ammeter reports that it is drawing 0.24 to 0.27 amps. When I plug that into to the keyboard port, the neuron flickers and the half-keyboard never lights up. 

I am inclined to believe that the KVM goes through some kind of initial discovery and configuration process, does not like the response it gets from the keyboard, and so resets the connection/keyboard and tries again. While I have a USB ammeter, I do not have a USB protocol analyzer so I cannot give you much more on this.

**Comment by jordansexton** [comment]
Possibly related: https://www.reddit.com/r/DygmaLab/comments/kp7rr9/is_the_raise_ps4_ps5_console_compatible/ (see comments)

The PS4 and PS5 also seem to require the ability to disable NKRO. Is there any way to do this currently?

**Comment by Nuru** [comment]
Any progress on this?

**Comment by AlexDygma** [comment]
We have now a list of compatible KMV's and are investigating the root cause for these issues, maybe it has something to do with packet segmentation on the usb HID peripheral, which is badly implemented on this version, or power packages which are completely lacking.

we will have to dig on this 😞.

**Comment by Nuru** [comment]
@AlexDygma Have you ruled out that it is a 6KRO vs NKRO issue?

**Comment by AlexDygma** [comment]
@Nuru Not really, but it doesn't seem so, as i don't have a KMV that fails the way yours does, i cannot test it, but i could make a branch with a modified version of the FW that only launches in 6KRO to clear this out.

**Comment by jordansexton** [comment]
@AlexDygma I would love to test that out to see if it resolves my issues too

**Comment by AlexDygma** [comment]
Ok!

I'll keep you posted!

**Comment by jordansexton** [comment]
Any update on this?

**Comment by Nuru** [comment]
@AlexDygma I would like to have some update on this before buying the tenting kit for a keyboard I still cannot use

**Comment by dmjeffmoore** [comment]
Any update on this? I also am having this issue but with the IOGEAR GUS4C2. Also, do you have a link to the compatible KVM list? Ty.

**Comment by Nuru** [comment]
@AlexDygma Last update on this was over a year ago. My brick is now 2 years old. Are you ever going to make it usable for me?

**Comment by jordansexton** [comment]
Also still waiting on this.

**Comment by gaborbernat** [comment]
I'd love this to be fixed too!

**Comment by mitchdz** [comment]
I'm seeing an issue with the TESmart B07D2ZMF5B 16x1 HDMI KVM, which I believe may be the same underlying issue as this thread. The situation is that the keyboard behaves as this post mentions when it's plugged into the controlling Mouse/Keyboard USB-A port, but works just fine when plugged into the generic USB2.0-A port. I have an ammeter as well and saw the same thing Nuru mentions.

This leads me to believe this issue occurs with KVMs which have keyboard shortcuts to control the KVM (e.g. change the channel, turn on auto-discovery, ect...) and when plugged into the controlling port.

**Comment by lorantfecske-bud** [comment]
In my case the problem wold be solved if I could switch to 6kro. It seems that the 1.0.9beta has some problems with switching nkro -> 6kro. I try to use the left ctrl - left shift - left alt + 6 press twice combination. The leds for those keys starts to blink, then no matter what I press, nothing seems to be going through or sometimes it's just sending 6 forever, however if I do the keycombo twice again, it just blinks again and then everything works again.

**Comment by herrbischoff** [comment]
> left ctrl - left shift - left alt + 6 press twice combination

Interesting, I've never heard or read about this combination. Also I'm on the same firmware but the combination does absolutely nothing for me. Customer support has been unresponsive for some time now, seemingly ever since Pedro is no longer a member of the Dygma team. Also not a good look that it's been ~20 months now since we heard from anyone from the development team.

**Comment by windowsrefund** [comment]
I was seriously thinking about buying this hardware until I read this thread.

**Comment by dmjeffmoore** [comment]
Instead of a KVM, I ended up using a USB switcher which works well - that may help some folk's use cases.

**Comment by mitchdz** [comment]
Yeah it works just fine with a regular USB switcher for me. Now I have a wireless keyboard (K830) plugged into my KVM controller port which I found out actually works better for my workflow anyways.

**Comment by AlexDygma** [comment]
Hey There people! sorry it took so long to get back to you all.

So referring to the 6KRO vs NKRO, firmware 0.3.5 solves that issue by fixing the switch command between the two modes, which is pressing the combo (all physical keys in their original positions)(LEFT CTRL+ LEFT ALT+ LEFT SHIFT+6) to switch back and forth, newer versions of the FW do not solve this issue due to lack of support of the superkeys on that FW. but if you can live without superkeys, it's much better.

In fact, we will be adding a way to use the v0.3.5 of the FW for the Raise with the latest version of Bazecor pretty soon, working on that right now (as the superkeys are still on beta and until after the launch of the Defy that won't change).

KVM switchers nowadays are crap!! as @dmjeffmoore and @mitchdz have suggested, it's way better to use a USB switch that is designed just with that in mind, either way, version 0.3.5 also addresses all of the issues that could affect KVM's stability like power signals and better package segmentation on the USB messaging, so give it a shot.

@windowsrefund, Im sorry to hear that but if you really need a KVM in your life, then custom or composite USB devices are not a thing for you 😞, it's a matter of poor implementation on the FW of those devices more than the keyboard itself, either way, using a USB Switch will solve those issues as they work as expected.

Best wishes!

**Comment by herrbischoff** [comment]
@AlexDygma Thanks for the update. My main issue is not necessarily a KVM switch but the macOS boot sequence. The early login screen when using File Vault does not respond to my Dygma Raise with firmware version 0.3.5, either in default state or after applying the key combination you mentioned. I notice that the combo appears to indeed modify its state but the computer still does not detect the Raise as a keyboard.

Additionally, when using newer firmware versions, my 11" iPad Pro does not properly recognise the Raise as a keyboard. With 0.3.5 it does.

So clearly, the "superkey" feature breaks lots of stuff all over the place. As nice as "superkeys" sound, I haven't found a use for them yet. In my opinion they're more trouble than they're worth. Gamers may see this differently. I'd be interested in hearing about what kind of feedback you've gotten — and whether a clear majority actually likes them.

For the longest time now, I have to attach a cheapo keyboard to even enter my password when restarting my machine. It's my only major gripe with the Raise but I find it to be quite unacceptable at this price point and overall build quality that a $3 keyboard works when a $300 one doesn't.

**Comment by Nuru** [comment]
@AlexDygma Thank you for the update. Unfortunately, I do not fully understand it. Are you saying it is too hard to make the latest firmware (with Super Key support) switch to 6KRO that you are putting it off indefinitely? That sounds hard to believe. Why can't this be a firmware setting managed by Bazecor?

@herrbischoff I use Super Keys because I have not found a better way to deal with the Raise's lack of an Escape key. 

More specifically, it comes with a key cap labeled "esc" but that key cap is on the key that normally types `` ` ~ `` on an ANSI keyboard. The normal Escape key is separated from the rest of the keys because it can be problematic if accidentally typed. I both need the Escape key and need to keep from accidentally hitting it, plus I need the `` ` `` key on a regular basis (10 times so far in this comment alone). So I use the key under the "esc" key cap for its ANSI use, but then where to I put Escape?

Like I said, I want Escape to be accessible, but hard to type accidentally. The best I have been able to come up with is to use Caps Lock as a Super Key, so that tapping it twice functions as "Escape". Everything else I have tried has an unacceptable error rate.

**Comment by GazHank** [comment]
@Nuru if the only aspect of superkeys that you are using is for the Escape key, I wonder if just moving Escape to a second layer would be an easy way to allow you to revert to the stable firmware.  That is of course assuming that there aren't other features of the beta that you need, and that the stable firmware actually enables the 6KRO mode switching that you need. 

That is to say use the escape key on the default layer to be ` and then Escape on layer 2?  much like I assume you would be using layers to access the Function Keys?

@AlexDygma are you able to share an more info on the specific plugin or settings that make 6KRO incompatible with the beta firmware?  I'd be keen to take a look, but don't want to repeat any investigations you have already completed.  I had initially thought that it could be an incompatability between the magicCombo and the DynamicSuperKeys plugins, but the magic combo does appear to be detected, it just doesn't switch the modes correctly... so I'm wondering if it is an issue with some of the clock frequency changes in the beta firmware creating an issue with the operation of USB-quicks plugin.  Perhaps the toggleKeyboardProtocol() "just" needs to be updated to account for the changes?

**Comment by herrbischoff** [comment]
@Nuru I have Escape mapped to Caps Lock. Its a largely useless key anyway and right on the home row. I'm an advanced Vim user. My second layer is reserved for mapping HJKL to arrow keys and function keys. Very convenient.

**Comment by GazHank** [comment]
@herrbischoff Do you know if it is 6kro or something else that is causing you issue on your iPad?  I've just checked my keyboard with the v1.0.9beta firmware on an iPad Pro running IPadOS 16.5 and it works without issues

**Comment by herrbischoff** [comment]
@GazHank Using the 6KRO key combo doesn't do anything for me. The only reliable way to use it with my iPad Pro (2018) is to use firmware 0.3.5. Otherwise, the Raise won't even get detected as a keyboard. Side effects include non-working keys (like Space) and persistent on-screen software keyboard. With 0.3.5 it works like with any other keyboard.

What still doesn't work in any configuration is entering my account password on my Mac at boot time when using File Vault WDE. It just won't react, 6KRO key combo or not. _Every single USB keyboard_ I've tried does. So this is unlikely to be some weird Apple-exclusive special case.

If I can provide you with more data, let me know what you need.

**Comment by erikbelusic** [comment]
> We have now a list of compatible KMV's and are investigating the root cause for these issues, maybe it has something to do with packet segmentation on the usb HID peripheral, which is badly implemented on this version, or power packages which are completely lacking.

I just got my raise, and was wondering where this list of compatible KVMs is - I tried plugging mine in and i end up hearing a high pitched noise and the keyboard is not responsive (but it IS lit up) - I assume the KVM I have doesn't provide enough power, and would like to switch to something that will work.

**Comment by alexpargon** [comment]
I am going to update this issue with some hardware that we know it works due to other users reporting the fact, in some cases and weird configurations it might not do so, but that should be rare.

- https://www.amazon.com/dp/B09BTSH2HW?ref=ppx_pop_mob_ap_share : Works, but whenever you connect another computer you’ll have to unplug and plug the Raise again to make it work-
- https://www.amazon.com/dp/B08H7YKJFT?ref=ppx_pop_mob_ap_share only supports two computers
- https://www.amazon.com/dp/B089ZW5PW5/ref=cm_sw_r_cp_api_glt_fabc_SPEYDXTVJA5SMEF2AGXX?_encoding=UTF8&psc=1 : works with up to four computers.
- https://www.amazon.com/Sabrent-Computers-Peripherals-Indicators-USB-SW30/dp/B074TYDJK2 : This also works on two computers
- https://smile.amazon.com/gp/product/B00DYP6I0E/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1 : Not exactly a KVM switch, but it works.

**Comment by mitchdz** [comment]
@alexpargon FYI the second last item you linked https://www.amazon.com/Sabrent-Computers-Peripherals-Indicators-USB-SW30/dp/B074TYDJK2%5B:%5D(https://www.amazon.com/Sabrent-Computers-Peripherals-Indicators-USB-SW30/dp/B074TYDJK2:) link is broken because it has :) at the end

**Comment by alexpargon** [comment]
> @alexpargon FYI the second last item you linked https://www.amazon.com/Sabrent-Computers-Peripherals-Indicators-USB-SW30/dp/B074TYDJK2%5B:%5D(https://www.amazon.com/Sabrent-Computers-Peripherals-Indicators-USB-SW30/dp/B074TYDJK2:) link is broken because it has :) at the end

Links fixed! thanks for the tip 😄, just copy-pasted them from document 😓

**Comment by Nuru** [comment]
@AlexDygma @alexpargon With the release of Bazecor 1.4.0 announced, do you have an update on this issue? 

@AlexDygma said almost a year ago now "we will be adding a way to use the v0.3.5 of the FW for the Raise with the latest version of Bazecor pretty soon, working on that right now (as the superkeys are still on beta and until after the launch of the Defy that won't change)."

So has there been progress made on this front? I have not seen anything related to it in the release notes or on this issue.

Regarding "KVM switchers nowadays are crap!!", certainly some (many?) are, but I have been very happy with the IOGEAR KVMs with this one exception. I confess their logo does not inspire confidence in me, but think I've been using their KVMs for about 20 years now and have been quite happy with them. That is part of why I do not want to give up on my IOGEAR KVM just to use the Raise. 

@herrbischoff Mapping the "caps lock" key to send "escape" was about the first thing I tried. I hit it by mistake several times a day, with significant consequences. (Most other keys, you  hit them by mistake, you can just backspace your way out of it, but with "escape" you tend to lose a lot of work with no "undo".)

**Comment by alexpargon** [comment]
> @AlexDygma @alexpargon With the release of Bazecor 1.4.0 announced, do you have an update on this issue?
> 
> @AlexDygma said almost a year ago now "we will be adding a way to use the v0.3.5 of the FW for the Raise with the latest version of Bazecor pretty soon, working on that right now (as the superkeys are still on beta and until after the launch of the Defy that won't change)."
> 
> So has there been progress made on this front? I have not seen anything related to it in the release notes or on this issue.
> 
> Regarding "KVM switchers nowadays are crap!!", certainly some (many?) are, but I have been very happy with the IOGEAR KVMs with this one exception. I confess their logo does not inspire confidence in me, but think I've been using their KVMs for about 20 years now and have been quite happy with them. That is part of why I do not want to give up on my IOGEAR KVM just to use the Raise.
> 
> @herrbischoff Mapping the "caps lock" key to send "escape" was about the first thing I tried. I hit it by mistake several times a day, with significant consequences. (Most other keys, you hit them by mistake, you can just backspace your way out of it, but with "escape" you tend to lose a lot of work with no "undo".)

Hey there @Nuru!

This is a FW Issue and as far as I know there is a specific version of the Defy firmware that solves it, the Raise has the v0.3.7 FW in which the swap works as expected.

Unfortunately, the Superkeys FW is still in beta and it doesn't have this issue addressed.

I'm really sorry for the inconvenience with this topic but there are alternatives, if you need further help I can try to push this issue to the FW team to see if they can replicate your exact issue.

**Comment by JayBusch** [comment]
I am also having great difficulty using the Raise as I would like.  I have it working with my Tessmart KVM via the USB port, which I can live with.  However I have a second workstation connected to the KVM via a wireless KVM extender.  That device will not recognize the Raise but has no issue recognizing my old Kinesis Freestyle 2.  I have tried flashing the firmware to 0.37 and using the 6KRO key combo.  When I execute the combo the keyboard flickers for a moment but never begins to work.

I know Dygma is focused on its next generation products, and in fact I've already signed up to reserve a Raise 2 (before I encountered this issue), but it seems these lingering issues with the Raise 1 are not being addressed.  We all payed a very heavy premium for a product for power users.  To be told we cannot use this keyboard in various scenarios when other much cheaper keyboards work just fine is pretty unacceptable.  I dont want multiple brands of keyboards in my workststations so my muscle memory has to.constantly adjust.  I want 2 Dygma Raises ideally, but I can't see myself buying the second if I'm told the first will simply not be supported, which will leave me to sell my Raise and buy two models of another brand.

Is there anything you can do to provide firmware that allows the keyboard to output the same signal as a typical, cheap, keyboard?  Am I to expect that if I buy a Raise 2 and there are some issues that I will be told the  same thing and not be able to use it like all other keyboards on the market?

Very frustrated here.  I love the hardware, but these firmware issues are intolerable frankly.  Please help keep me a Dygma fan and customer.

**Comment by JayBusch** [comment]
Seems like the UHK team has a handle on this.

https://github.com/UltimateHackingKeyboard/firmware/issues/536

**Comment by wez** [comment]
FWIW, I have Defy and am intensely disappointed that it does not function at all when connected to my expensive but totally excellent [Level1 KVM](https://www.store.level1techs.com/products/p/14-display-port-kvm-single-4computer-6zepx-s4kf3).  The Defy is the only device I own that does not work with this KVM. I've built my own custom keyboards with varying levels of sketchy firmware and those all work with the KVM.

I think it is disingenuous to blame this incompatibility on KVMs being crap when it is only Dyma's devices that have this issue.

**Comment by herrbischoff** [comment]
@JayBusch and @wez, at this point in time – after years of deflection, silence and close to zero firmware updates – I'm forced to come to the conclusion that Dygma does not have a single C++/embedded developer on staff, much less one that actually understands how to fix the issue. There are JavaScript developers working on Bazecor though.

From the outside looking in, it appears that they have taken some FOSS firmware, hacked about it a bit, managed to work themselves into a cul de sac and finally called it a day. Either that or the team has decided simply not to care. This may not be accurate but in any case, they're doing a bad job communicating their priorities by not communicating in any meaningful way.

While the build quality is still great, a $400+ keyboard which cannot do the basic tasks a $5 keyboard can is unacceptable, period. That it cannot do it because of a firmware issue is almost ridiculous. As things stand now, I will actively recommend against Dygma’s products when asked and certainly not buy anything from them again.

Starting to sell follow-up products to ones whose issues are numerous and unresolved is a move that erodes trust. To call me disappointed is an understatement.

**Comment by jordaaash** [comment]
> Starting to sell follow-up products to ones whose issues are numerous and unresolved is a move that erodes trust. To call me disappointed is an understatement.

Agreed, I'm also extremely disappointed by the lack of action on this over the course of years while Dyma has launched, and is still launching, multiple new products. I'm [unable to use the Raise with my PS5](https://github.com/Dygmalab/Bazecor/issues/159) because of this, and I will not purchase another Dygma product until it's fixed.

**Comment by phming** [comment]
I'm not sure if I would have bought a Defy had I seen this discussion earlier. This discussion is missing updates from Dygma. Is Dygma at all working on this issue? What is the current status?

Anyhow, I recently got my Defy in the mail and started playing around with Bazecor on Windows, even though I use FreeBSD as my main OS (I expected Bazecor wouldn't be readily available on there, so far so good). What I didn't expect was that the keyboard wouldn't work at all; well, funnily enough it works in the boot loader stage but not after login is available. So after I found this discussion I was about to get quite mad and was almost ready to get ready to refund the whole thing. (spoiler I am now typing this on my Defy connected to my FreeBSD PC connected over my monitor's USB3 hub indeed).

While I cannot speak to the KVM issue, PS5 and MacOS have one thing in common (afaik) and that is that they are both based on or rather use parts of (Free)BSD.

Now thankfully, before getting really mad that my shiny new 550 CHF keyboard could not be used with my main operating system, I discovered a fix: 
simply add 
`hw.usb.usbhid.enable=1` to /boot/loader.conf and reboot. After that it's now working as expected. I really hope Bazecor will eventually be available on this OS as well, but thankfully that is not my main concern.  

How does one implement this fix on MacOS or PS5? I don't know but it's probably easier on the Mac than on PS5 and it's not certain that it will work. But I figured that maybe this could help give devs a hint. Regarding Dygma I expected to see better, after all the product design convinced me. Now I'd like to see at least better communication. Is it being worked on or are customers left standing in the rain?

greetings from switzerland

**Comment by MGousseff** [comment]
Hello @AlexDygma : Is there a fix yet ? It seems the issue is 4 year old. I have a Dygma Raise that I don't use on a KVM, but was contemplating buying a raise 2 for a KVM setup ?

**Comment by herrbischoff** [comment]
@MGousseff FWIW, I've had a written exchange with support some time ago. The Dygma team has decided that solving this issue is not a priority for them. There is now a 6KRO firmware variant to flash in Bazecor but it didn't solve any of my issues. In fact, it contains bugs that freeze the keyboard on wake.

Like I'd written before, Dygma clearly has no experienced embedded developer on staff who understands in depth the free firmware they've forked and hacked on.

I am considering returning my Raise for a refund since it's been defective in this regard since its initial purchase.

**Comment by Belpheg0rsPrime** [comment]
Just wanted to lend my voice and say that I have been perfectly happy for years with my Raise until I also bought a [level1techs KM switch](https://www.store.level1techs.com/products/p/4-port-km-switch-with-usb-32-gen-1-mouse-roaming-function). This was a major upgrade for me from a previous USB 3.0 switch that didn't provide enough power to the raise (causing the LEDs to subtly flicker) and that was extremely slow to switch between PCs ([link](https://www.amazon.com/gp/product/B07FQT43DM/ref=ppx_yo_dt_b_search_asin_title?ie=UTF8&psc=1) for the curious). I am honestly shocked to read this issue and see the current state of it. It has certainly warned me off the Raise 2. L1T has the nicest KM switch I have ever seen, and I certainly did my research. To see that the Raise can't function as a proper HID device with it is extremely disappointing. I am hoping for an update from @alexpargon or @AlexDygma but I'm not holding my breath.

**Comment by erikbelusic** [comment]
No updates or direction in 6-12 months =/

**Comment by johmsalas** [comment]
4 years and a half later. Still no solution

**Comment by Jertusz** [comment]
Adding my 2 cents.

I also experience the exact same issue that's mentioned in this thread, for the past 2 years I've mentioned it a few times in the official Dygma discord, and the response I got from `#minues` aka Dygman, was the following:

> First, I would check that the issue that you're having is only through the KVM. 
> Then, the easiest solution is to try a different KVM. Generally, they work, it's just some that don't.

I find this approach ridiculous and given how many people experience this issue in this thread and discord as well, very offending. I don't think that just ignoring this will make it go away @alexpargon @AlexDygma 

I'd expect some kind of definite information whether or not you are actually doing anything to resolve this issue, and if not some kind of return policy based on that would be nice.

**Comment by johmsalas** [comment]
Seems like the only solution is to expose this issue in public channels, which unfortunately will affect the brand, but given the cost of this keyboards, premium support is expected. 

@alexpargon do you know who could own this issue. Dygma Support, advise is required here

**Comment by gaborbernat** [comment]
I don't think that's how this works. Just because something has a premium price (arguable this one considering the feature set provided) doesn't mean it has to do everything imaginable. Unless they didn't advertise as supporting your KVM your case would be very thin (and I don't think they did). You're implying malice somewhere I don't think there's one. 

Granted some acknowledgement from Dygma would be nice, but likely they would only get back here once they have time and resources to look into this.

**Comment by Nuru** [comment]
@gaborbernat wrote:
> Granted some acknowledgement from Dygma would be nice, but likely they would only get back here once they have time and resources to look into this.

Dygma acknowledged the problem over 5 years ago, on January 2, 2020, and promised a fix "with the next firmware update". They haven't even fixed it **with the next hardware model release**, let alone the next firmware update. 

Over a year ago they were pointed to another project that fixed the problem in 2 ways:

1. [Removing NKRO support](https://github.com/UltimateHackingKeyboard/firmware/issues/536#issuecomment-1291550885) and reverting to 6KRO, as originally diagnosed by Dygma in 2019.
2. [Rearranging the order of USB Interfaces advertised](https://github.com/UltimateHackingKeyboard/firmware/issues/287)

These are not the kinds of changes that should be hard for a keyboard manufacturer to make. If they have had time and resources to design 2 more keyboards (Defy and Raise 2), they should have had time and resources to fix this. To blame the KVMs is lame, especially since the KVMs that the Dygma keyboards are failing are some of the best ones on the market.

**Comment by jordaaash** [comment]
Seems to me like they muted this one a long time ago. Perhaps a linked issue should be opened against the firmware repo? Maybe this isn’t a “Bazecor” issue but it is a firmware issue, and it really should have been fixed like Dygma said it would be years ago. I still can’t use it with consoles as a result.

**Comment by Jertusz** [comment]
@jordaaash, I don't think it's necessary to open a different issue just yet. I've talked with `#minues` aka Dygman on discord on Saturday again, and he promised me that he'll reply here today. I guess if there is no reply by EOD today, then opening another issue might be worth a shot, because as he stated:

>Please accept my apologies because keeping track of old GitHub issues is difficult. I check every day the new ones and get notified on replies for the ones I've answered, but not for older ones.

Which doesn't really explain why @alexpargon didn't escalate the issue to him, but that's besides the point.

**Comment by MGousseff** [comment]
Oh, I thought I was being useful and created an issue this morning on the raise firmware repository. https://github.com/Dygmalab/Raise-Firmware/issues/34
I thought it was better to make an issue in the repo it was in the end the most likely to be useful after reading the comments here and the problem probably not being a bazecor problem.
Do you think I shall delete it ?

**Comment by MiquelDygma** [comment]
Hi!

First of all, my apologies for not replying to this thread sooner. I get notified of new issues and issues I have commented on, but not older issues with new comments. I should have been checking issues with new comments manually, but I didn't, so that's totally on me. I'll add that to my daily checks so it doesn't happen again. 

For the Dygma Raise, we released a firmware version with 6KRO (0.3.7 - 6KRO), but that hasn't solved the issue with some KVMs. 

There's clearly some incompatibility with the Raise firmware (and the Defy and Raise 2, too) with some KVMs, but our efforts so far have not been successful. Since it affects only a small part of the users and has a relatively easy solution (other KVMs work), we've had to prioritize other issues. 

The dev team is still working on connectivity improvements for the Defy and the Raise 2, and if we find the issue and the solution, we'll implement it on the Raise, but we can't promise anything. 

We've added a list of KVMs and USB switches that work in this link: https://support.dygma.com/hc/en-us/articles/22668824507421-Compatibility-with-USB-hubs-and-KVM-switches 

Some users have also reported success using the ports on their KVMs that are NOT keyboard and mouse ports.

We've also added information regarding KVM and USB switch compatibility to the FAQs on our website. 

@MGousseff I'll paste that reply there, but most people use this repository both for Bazecor and Firmware issues.

Once again, my apologies for the late reply and for the fact that we haven't been able to make the keyboard fully compatible with all KVMs. 

Best,
Miquel/Dygman

**Comment by MGousseff** [comment]
@MiquelDygma Thank you for your answer. If you think the other issue should be deleted, fine by me, I just wanted to be sure the right part of your team had the right information. 
I'll check the list of compatible KVMs. Thank you.

**Comment by Jertusz** [comment]
First of all, thank you for responding @MiquelDygma, that's definitely a step in the right direction.

>Since it affects only a small part of the users and has a relatively easy solution (other KVMs work), we've had to prioritize other issues.

This is understandable, however, assuming that most of the participants here have this issue, and many people reported it directly on Discord instead of participating here, the issue doesn't affect 2-5 users, it affects dozens. I was lucky enough to buy the OG Raise for a much lower price (when there was no info available about the KVM issues), but given that the cheapest variant of Raise 2 starts at $369 and goes into $600-700 range, saying _other KVMs work_ is a bit of a stretch. So on top of buying a **very** expensive keyboard we're expected to switch out our KVMs, because you decided that users with KVMs are secondary to the ones without them.

>The dev team is still working on connectivity improvements for the Defy and the Raise 2, and if we find the issue and the solution, we'll implement it on the Raise, but we can't promise anything.

Working on improvements for Defy and Raise 2 is fully understandable, but whatever the improvements are, I doubt users affected by this issue will enjoy any of them if their keyboards don't work. On top of that, this is the same response I got 2 years ago, only then it was phrased a bit differently, and so far, nothing has happened. Are you **actively** looking for a solution for this issue, or if you just accidentally stumble upon one then you'll implement it?

>Some users have also reported success using the ports on their KVMs that are NOT keyboard and mouse ports.

Doesn't this point directly to an issue with your software? If some users can use the keyboard in a regular port, not the HID one, then I think it's safe to assume, that your software has issues with producing valid HID for the device. FYI, kaleidoscope, which you've forked to create the firmware has issues that have been merged, that might help on that: https://github.com/keyboardio/Kaleidoscope/issues/1305

>We've also added information regarding KVM and USB switch compatibility to the FAQs on our website.

To be clear, this was added yesterday or today, and only after being suggested by me. It wasn't your initiative to be more transparent about that. But it's great that you updated that.

**Comment by MiquelDygma** [comment]
Let me address some of your comments

> So on top of buying a very expensive keyboard we're expected to switch out our KVMs, because you decided that users with KVMs are secondary to the ones without them.

It's not exactly like that; there are issues that affect more people and we're addressing those first. 

>  Are you actively looking for a solution for this issue, or if you just accidentally stumble upon one then you'll implement it?

We're working on improving the HID communication of the neuron with the computer (among other things), and we expect that to solve this issue and other issues.

> Doesn't this point directly to an issue with your software? If some users can use the keyboard in a regular port, not the HID one, then I think it's safe to assume, that your software has issues with producing valid HID for the device. FYI, kaleidoscope, which you've forked to create the firmware has issues that have been merged, that might help on that: https://github.com/keyboardio/Kaleidoscope/issues/1305

Yes, as I said, "There's clearly some incompatibility with the Raise firmware (and the Defy and Raise 2, too) with some KVMs". But it's not the key rollover, as you mentioned that the 0.3.7 6KRO firmware didn't solve the issue with your Raise. 

> To be clear, this was added yesterday or today, and only after being suggested by me. It wasn't your initiative to be more transparent about that. But it's great that you updated that.

That is correct; I added it this morning since you mentioned it wasn't there.

**Comment by johmsalas** [comment]
@MiquelDygma Thank you for paying attention to this issue. As a short-term solution, I'm going to invest in one of the KVMs mentioned. I understand your priorities and would appreciate, if not support for every KVM, at least accountability for some references

I believe the market for KVMs will grow soon. There’s considerable friction due to the recent shift to remote work. The cost of switching contexts is too high when moving from the company's computer to the home setup. KVMs help ease this friction. It seems there are two markets: on one side, custom keyboards, and on the other, individuals struggling with different computer setups (and its overwhelming friction). It’s only a matter of time before both recognize each other's presence

**Comment by jordaaash** [comment]
This also causes the Raise (and Raise 2) not to work with gaming consoles, such as the PS5. This probably affects more users than expected. It sounds like maybe [6KRO support fixes this](https://github.com/UltimateHackingKeyboard/firmware/issues/536#issuecomment-2513760055) for some people. Is there a way this could be added to current firmware via a setting in Bazecor?

**Comment by MiquelDygma** [comment]
@johmsalas than you. My short-term goal is to keep documenting KVMs that work and those that have documented issues. The long-term goal is to maximize compatibility, although sometimes things are out of our control. For example, some KMVs have issues with keyboards that have mouse keys in the firmware (they confuse them with mice) but work when connected to the auxiliary port.

@jordaaash, we released a 6KRO version of the Raise firmware, and we want to do the same for the Raise 2 and the Defy. It's in the backlog but not actively being developed at the moment.

**Comment by Jertusz** [comment]
@MiquelDygma Could you share a bit more about the Kaleidoscope dependency update process? From what I saw, even though you're using Kaleidoscope, the update frequency is rather small, or even non-existent for the past 2 years. At least, that's what the commits suggest here: https://github.com/Dygmalab/Raise-Firmware/commits/master/. Perhaps bumping the Kaleidoscope version would fix the issues mentioned here? Given that the fix for HID devices within Kaleidoscope, that I mentioned earlier, was implemented after 3.6.0 was released.

**Comment by MiquelDygma** [comment]
@Jertusz, that merge was done for the Dygma Raise in the 0.3.7 firmware, and we also released the 0.3.7 - 6KRO version. 

That solves some issues, but not all, as you reported the 0.3.7 firmware not working with your KVM. 

We tried for half a year to merge the latest kaleidoscope with the 1.0.12 branch, but we couldn't make Superkeys work when merging upstream (that was our development), so we decided to continue developing on our fork without merging.

**Comment by Jertusz** [comment]
@MiquelDygma, you're right, 0.3.7 is the latest version, my mistake.

Would you be able to share a FW version (that you built during development or something), or point me to some resources I can use to flash my Raise with the latest Kaleidoscope version? I can live without the Superkeys, so would love to try flashing the Raise with the updated FW containing latest Kaleidoscope changes, to see if that may help resolve the issue.

**Comment by MiquelDygma** [comment]
@Jertusz, the last time we merged upstream was the 0.3.5 firmware, which included the HID improvements (https://github.com/keyboardio/Kaleidoscope/issues/1305). Then we made some fixes and released 0.3.7 

You can see the Raise firmware repository here in case you want to clone that repository and merge upstream with Kaleidoscope:
https://github.com/Dygmalab/Raise-Firmware

But the HID improvements that you mentioned are implemented in the 0.3.7 firmware.

**Comment by nvonbenken** [comment]
I have a different KVM - from Level1Techs - and have all the same issues here with my Defy to the point that I've stopped using it. It is bizarre to me that a keyboard just doesn't work on an HID port. 

Are the HID improvements mentioned above available for the Defy as well or just for the Raise?

**Comment by MiquelDygma** [comment]
@nvonbenken, if you tell me which model you have, I can add it to our list of incompatibilities. 

As I explained before, the Dygma Defy and Raise 2 are built on top of the 1.0 version of the Raise firmware, which was not merged upstream with the Kaleidoscope update with HID improvements. This is because doing that merge broke the superkeys completely. 

Nonetheless, those HID improvements don't solve the problem for everyone. Some KMVs have issues with keyboards that have mouse keys in the firmware (they confuse them with mice) but work when connected to the auxiliary port.

**Comment by nvonbenken** [comment]
@MiquelDygma  I have this one here: https://www.store.level1techs.com/products/p/14-kvm-switch-triple-monitor-2computer-pne46-fxlmw

**Comment by MiquelDygma** [comment]
@nvonbenken, thank you! Have you tried connecting the Defy to a non-HID port of the KVM?

**Comment by nvonbenken** [comment]
@MiquelDygma  Yes I have and it works fine for the most part. 

There are instances where it will lock up and not work entirely. Usually unplugging it and plugging it back in resolves this.

The other issue is that I cannot use keybinds to switch inputs on my KVM without it being on the HID port.

**Comment by MiquelDygma** [comment]
@nvonbenken I see. Yes, that's part of the limitation of having to use the HID ports. We're going to re-work the neuron communications during the following months, but there's a change that won't fix your issue with the HID port.

**Comment by nvonbenken** [comment]
Any updates to this with the beta firmware released today?

**Comment by MiquelDygma** [comment]
The beta firmware improves the behavior of the keyboard while wired, so it could improve how it behaves while switching devices with a KVM. But the problem with not being able to use the HID port is something that's caused by the KVM being confused by the mouse keys of the keyboard. 

PS: the new beta firmware is for the Defy and the Raise2, not the OG Raise.

**Comment by nvonbenken** [comment]
I'll see about trying it out this weekend.

I do have the Defy but it exhibits the same issues as described here. Not sure if a separate issue should be opened or this one be made more generic to cover the Defy.

**Comment by MiquelDygma** [comment]
Maybe it would be better if you open a new one for the Defy once you test it.
